### PR TITLE
RD-1751 Add basic tests for Deployments View widget

### PR DESCRIPTION
**Pull request by Gelio**
The PR adds a few basic tests for the Deployments View widget.

Don't be afraid by the number of lines added :smile: Most of them come from the test fixture for deployments

All passed locally:
image

## Changes in this PR

1. Migrate the sites commands to TypeScript

    Sites commands are used in the new test suite.

2. Basic tests:

    1. The created deployment is displayed in the table
    2. Hiding and showing columns (`fieldsToShow` in the widget configuration) is working correctly
    3. The deployment status icon is displayed correctly and has a popup

    I have tried to make the tests rely on accessible properties (`aria-label`, `role`, text), rather than classes. I think I managed to do that well, except for a few cases :slightly_smiling_face: 

    The added fixture will make it easier to test functionalities such as displaying the number and statuses of subdeployments, and the latest execution progress. Those will be implemented and tested in future tickets.

3. Add handling `process.env.TEST` (set to `'1'` during e2e tests)

    Necessary to get `deploymentsViewWidget` to render during system tests.

    I was hoping to set `process.env.NODE_ENV` to `test` during e2e tests, but webpack overrides `process.env.NODE_ENV` based on the `mode` it runs in. Thus, in e2e tests it always has `process.env.NODE_ENV === 'production'` :worried: 

    To verify that it does not break production, I ran the production build locally and here's the value of `process.env.TEST`:
    image

## System tests

https://jenkins.cloudify.co/job/Stage-UI-System-Test/315/

Ran 2021-03-23 10:56 CET

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
   "scripts": {
     "audit": "npm audit --production && cd backend && npm run audit",
     "beforebuild": "npm ci --prefer-offline && (cd backend && npm ci --prefer-offline)",
-    "build": "cross-env NODE_ENV=production webpack --config webpack.config.js --mode=production --bail",
+    "build": "webpack --config webpack.config.js --mode=production --bail",
```
Comment (Gelio) [inline package.json:20]: See message of 509afa4aebe2ebba781ec4c8a7b308abe55fc283 for more details on this change

File: test/cypress/integration/widgets/deployments_view_spec.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+            cy.contains('hello-world-one')
+                .parents('tr')
+                .find('i.spinner[aria-label="In progress"]')
+                .as('inProgressIcon')
+                .trigger('mouseover');
+            cy.root().parents('body').find('.popup').contains('In progress');
+            cy.get('@inProgressIcon').trigger('mouseout');
+
+            cy.contains('one-in-warsaw').parents('tr').find('td:nth-child(1) i').should('not.exist');
```
Comment (Gelio) [inline test/cypress/integration/widgets/deployments_view_spec.ts:102]: I could not find a better way to check that the icon does not exist

File: test/cypress/support/commands.ts [hunk 3]
```diff
@@ -176,13 +176,14 @@ const commands = {
         }
         cy.waitUntilPageLoaded();
     },
-    usePageMock: (widgetIds?: string | string[], widgetConfiguration: any = {}) => {
+    usePageMock: (
+        widgetIds?: string | string[],
+        widgetConfiguration: any = {},
+        additionalWidgetIdsToLoad: string[] = []
```
Comment (Gelio) [inline test/cypress/support/commands.ts:182]: Allows using widgets that are not displayed initially in the tests

**Comment by Gelio** [comment]
It seems that the changes around `NODE_ENV` do not work, since webpack overrides `NODE_ENV` to `production` or `development` when the `mode` property is set. Thus, `NODE_ENV` will always be set to `production` in e2e tests.

I'm working on a different solution, will push it out in 1h

File: widgets/deploymentsView/src/widget.tsx [hunk 4]
```diff
@@ -24,8 +24,8 @@ interface DeploymentsViewWidgetConfiguration {
 
 const i18nPrefix = 'widgets.deploymentsView';
 
-// TODO(RD-1224): remove environment check
-if (process.env.NODE_ENV === 'development') {
+// TODO(RD-1226): remove environment check
+if (process.env.NODE_ENV === 'development' || process.env.TEST) {
```
Comment (Gelio) [inline widgets/deploymentsView/src/widget.tsx:28]: `process.env.TEST` is undefined in production:

image

And `process.env.NODE_ENV === 'development'` is substituted to `true` in development:
image

**Comment by Gelio** [comment]
Turns out `process.env.TEST` is still `undefined` during test. `EnvironmentPlugin` does not forward it. I need to investigate

**Comment by qooban** [comment]
@Gelio Try using `supportedEditions` field on widget definition. You can just pass something like `['my-super-edition']` and the widget should not be presented to the users.

**Comment by Gelio** [comment]
> @Gelio Try using `supportedEditions` field on widget definition. You can just pass something like `['my-super-edition']` and the widget should not be presented to the users.

Unfortunately, that's not what I was looking for :sweat_smile: I am already disabling the widget by not calling `Stage.defineWidget` in production. What I was looking for is a way to know whether we are in actual production, or in e2e tests :slightly_smiling_face: I ended up adding another `process.env.TEST` variable to determine it

---

Turns out it was not working because `EnvironmentPlugin` was not being applied at all :slightly_smiling_face: It was inserted (by me :smile: ) in the wrong place some time ago.

File: webpack.config.js [hunk 5]
```diff
@@ -216,9 +217,9 @@ module.exports = (env, argv) => {
                             }
                         }
                     }),
+                    environmentPlugin,
                     isProduction && getProductionPlugins(env && env.analyse === 'main')
-                ]),
-                environmentPlugin
```
Comment (Gelio) [inline webpack.config.js:221]: It was placed outside of the array :sweat_smile:

File: test/cypress/integration/widgets/deployments_view_spec.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    const blueprintName = 'deployments_view_test_blueprint';
+    const deploymentName = 'deployments_view_test_deployment';
+    const siteName = 'Olsztyn';
+    const blueprintUrl =
+        'https://github.com/cloudify-community/blueprint-examples/releases/download/latest/simple-hello-world-example.zip';
+
+    before(() => {
+        cy.activate()
+            .deleteSites(siteName)
+            .deleteDeployments(deploymentName, true)
```
Comment (kubama) [inline test/cypress/integration/widgets/deployments_view_spec.ts:14]: Not sure if it matters in this particular case, but since it's a deployment that is assigned a site, not the other way around, I think deployments should be deleted first, then sites.

File: test/cypress/integration/widgets/deployments_view_spec.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            .uploadBlueprint(blueprintUrl, blueprintName)
+            .deployBlueprint(blueprintName, deploymentName, { webserver_port: 9123 })
+            .createSite({ name: siteName })
+            .setSite(deploymentName, siteName);
+    });
+
+    const useDeploymentsViewWidget = (routeHandler?: RouteHandler) => {
+        cy.interceptSp('GET', 'deployments', routeHandler).as('deployments');
+        // NOTE: add widget through Edit Mode to apply default configuration automatically
+        cy.usePageMock([], {}, [widgetId]).mockLogin().addWidget(widgetId);
```
Comment (kubama) [inline test/cypress/integration/widgets/deployments_view_spec.ts:25]: Is there any reason why we want the default configuration? Using specified configuration would make the test faster, and speed was the actual reason behind introducing `usePageMock` allowing any widget to be added to the test page out of the box, without the need for Edit Mode.

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 6 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts:14]: Good point :slightly_smiling_face: Copy-paste didn't work well :smile:

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 7 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts:25]: I understand the concern about the speed of the test. In this case, I decided that the pros outweigh the cons

The default configuration is quite lengthy (it has the list of columns visible in the table). I did not want to duplicate it and hardcode it here, because:
1. The test gives less confidence that the functionality works
2. It makes the test not pick up changes to the default widget configuration
3. That would tie the test to the implementation, which I did not want to do, as I prefer black-box testing (changing the field name in the table would require changing the tests)

---

As I see it, the problem with the widget configuration lies in **when** the default configuration is applied. As I understand, applying default configuration happens when the page is created or the widget is added to the page, and not when it's rendered. Thus, when using `usePageMock` the configuration is empty, the widget does not merge the default configuration into the passed `configuration`, but just uses the empty config, which makes the widget fail.

What I would suggest as a global fix would be to use the widget's default configuration when the widget is being rendered and its configuration is missing. Then, I would happily render the widget with the empty config here, and it would use the default one.

What do you think about such a solution? If you're ok, I will create a tech debt item to introduce such changes in the future

Comment (kubama) [reply test/cypress/integration/widgets/deployments_view_spec.ts:25]: I'm all for the proposed solution.

However, since it is unlikely that the tech debt item will be implemented any soon I would still consider using specified configuration. The points that you made are general and apply to any widget test, not only to this particular one, so if we agreed on them then we should start moving all widget tests to this approach. As it would increase tests execution time considerably it would be an opposite direction to what we were trying to achieve so far (system tests execution time was an issue and `usePageMock` was introduced to address it).

File: package.json [hunk 8]
```diff
@@ -34,7 +34,7 @@
     "docWidgets": "node scripts/updateReadmes.js",
     "e2e": "cypress run -C node_modules/cloudify-ui-common/cypress/cypress.json",
     "e2e:open": "cypress open -C node_modules/cloudify-ui-common/cypress/cypress.json",
-    "e2e:ci": "npm run beforebuild && run-system-tests ${TEST_SPEC}",
+    "e2e:ci": "npm run beforebuild && cross-env TEST=1 run-system-tests ${TEST_SPEC}",
```
Comment (qooban) [inline package.json:37]: Hmm... We are adding new environmental variable to be able to include specific widget in the test-build. Not sure I like this solution as I see it is temporary or we plan to use it more extensively?

Maybe in `widgets/deploymentsView/src/widget.tsx` instead of `TEST` env-var we could check some env-var available only in the Jenkins CI? Loudly thinking, I'm not sure it would be the best.

**Review by qooban — COMMENTED** [review]
In general looks good 👍🏻 

However, I'm not sure about introducing new env-var for that specific test.

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 7 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts:25]: Solid points, I will refactor the test to use the hardcoded version of the configuration and create tickets to introduce the solution I described earlier :slightly_smiling_face:

File: package.json (hunk 8 above)
Comment (Gelio) [reply package.json:37]: > Not sure I like this solution [...]

Yeah, I was not too sure about it either, but I did not come up with a better way. I was trying to use `NODE_ENV = 'test'`, but not only that did not work, that would also alter the bundle (since other packages, e.g. `React`, strip away some code in production, meaning that the behavior in tests could be a bit different than in production). Thus, I decided that a new variable would be a good enough solution

> [...] or we plan to use it more extensively?

My intent is to be able to show and test widgets, as well as enable some functionality only in test and development environments, but not in production. This applies to Environments, and probably future features I will be working on :slightly_smiling_face: I see benefits to this approach (I call it "poor man's feature flag") compared to feature branches.

That said, I didn't come up with a better solution for now. Whether this solution stays, or we implement something else, is an open question. My only aim is what I described in the paragraph above :smile: 

> Maybe in widgets/deploymentsView/src/widget.tsx instead of TEST env-var we could check some env-var available only in the Jenkins CI?

Keep in mind that the list of environment variables that are available in the code has to be listed in the `EnvironmentPlugin` in webpack's config. This means that if we want to use some Jenkins environment variable, we would have to specify it there, along with a default value. It's not a large overhead, but if we use an environment variable just to detect the environment, I would prefer a service-agnostic `TEST` to some obscure `JENKINS_...` :smile: 

Also, if we run `npm run e2e:ci` locally, then the Jenkins variable will not be set, meaning that the code being tested will be different than the one on Jenkins :smile: To fix it, we would have to define that variable ourselves, probably in the `npm run e2e:ci` script. Then, it would not solve the problem of having to define an env variable anyway :slightly_smiling_face: 

To sum up, I don't see a better way, and, in my view, a service-agnostic `TEST` variable is superior to something Jenkins-related :slightly_smiling_face:

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 7 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts]: The execution time when running tests locally went down 20% (from 55 seconds to 44 seconds) :slightly_smiling_face: 

Here are the tickets:
1. https://cloudifysource.atlassian.net/browse/RD-1820
2. https://cloudifysource.atlassian.net/browse/RD-1821

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 6 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts]: Done :slightly_smiling_face:

**Review by kubama — APPROVED** [review]

File: test/cypress/integration/widgets/deployments_view_spec.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        it('should allow changing displayed columns', () => {
+            useDeploymentsViewWidget();
+
+            cy.log('Hide some columns');
+            cy.editWidgetConfiguration(widgetId, () => {
+                getFieldsDropdown().within(() => {
+                    cy.contains('Blueprint Name').find('.delete.icon').click();
+                    cy.contains('Location').find('.delete.icon').click();
+                });
+            });
```
Comment (kubama) [inline test/cypress/integration/widgets/deployments_view_spec.ts:52]: Actually, this could be part of the initial widget configuration.

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (kubama) [reply test/cypress/support/commands.ts:182]: Is this still needed?

File: package.json (hunk 8 above)
Comment (qooban) [reply package.json:37]: OK, got it, thanks for detailed explanation. 

I think that maybe we can go further and have that variable turned on in `run-system-tests` script. But for now we can leave it as it is.

**Review by qooban — DISMISSED** [review]

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 9 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts:52]: Yeah, very much so. Would it be ok if I leave it be as-is for now? :smile: I'd like to continue working on other tasks that depend on this PR and reduce the "branch-chaining"

**Comment by Gelio** [comment]
I've just queued up the system tests for this branch: https://jenkins.cloudify.co/job/Stage-UI-System-Test/315/

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 9 above)
Comment (kubama) [reply test/cypress/integration/widgets/deployments_view_spec.ts:52]: Wouldn't it be a 5-10 minutes change? Adding an optional parameter to `useDeploymentsViewWidget` allowing to override `fieldsToShow`.

File: package.json (hunk 8 above)
Comment (Gelio) [reply package.json:37]: > I think that maybe we can go further and have that variable turned on in run-system-tests script. But for now we can leave it as it is.

Yup, that would make sense :+1:

File: test/cypress/integration/widgets/deployments_view_spec.ts (hunk 9 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployments_view_spec.ts:52]: Yes, it would. I did not want to context-switch to introduce a change I thought was minor and could be optional :slightly_smiling_face:

**Comment by Gelio** [comment]
I've added more more commit to address @kubama's comment https://github.com/cloudify-cosmo/cloudify-stage/pull/1253#discussion_r599398075

**Review by qooban — APPROVED** [review]

**Comment by Gelio** [comment]
The 2 system tests that failed are:
1. Known `topology_spec`
2. A not-updated test for `deployment_actions_spec.js` after #1256  (which I've reported on slack)

Both are unrelated to this PR, so I'm merging it

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (Gelio) [reply test/cypress/support/commands.ts:182]: Sorry, I did not notice your comment earlier. It's not leveraged in this PR, but will be used in a future PR once the details pane is implemented for this widget. We need to load widgets that are rendered in that details pane, which are not rendered in the top level
